_bleio/adapter_.py

```diff
@@ -0,0 +1,222 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`_bleio.adapter_`
+=======================================================================
+
+_bleio implementation for Adafruit_Blinka_bleio
+
+* Author(s): Dan Halbert for Adafruit Industries
+"""
+from __future__ import annotations
+from typing import Iterable, Union
+
+import asyncio
+import platform
+import threading
+import time
+
+from bleak import BleakClient, BleakScanner
+
+from _bleio.address import Address
+from _bleio.connection import Connection
+from _bleio.exceptions import BluetoothError
+from _bleio.scan_entry import ScanEntry
+
+Buf = Union[bytes, bytearray, memoryview]
+
+# Singleton _bleio.adapter is defined at the ned of this file.
+adapter = None  # pylint: disable=invalid-name
+
+
+class Adapter:
+    # Do blocking scans in chunks of this interval.
+    _SCAN_INTERVAL = 0.25
+
+    def __init__(self):
+        if adapter:
+            raise RuntimeError("Use the singleton _bleio.adapter")
+        self._name = platform.node()
+        # Unbounded FIFO for scan results
+        self._scanning_in_progress = False
+        self._connections = []
+        self._bleak_loop = None
+        self._bleak_thread = threading.Thread(target=self._run_bleak_loop)
+        # Discard thread quietly on exit.
+        self._bleak_thread.daemon = True
+        self._bleak_thread_ready = threading.Event()
+        self._bleak_thread.start()
+        # Wait for thread to start.
+        self._bleak_thread_ready.wait()
+
+    def _run_bleak_loop(self):
+        self._bleak_loop = asyncio.new_event_loop()
+        # Event loop is now available.
+        self._bleak_thread_ready.set()
+        self._bleak_loop.run_forever()
+
+    def await_bleak(self, coro, timeout=None):
+        """Call an async routine in the bleak thread from sync code, and await its result."""
+        # This is a concurrent.Future.
+        future = asyncio.run_coroutine_threadsafe(coro, self._bleak_loop)
+        return future.result(timeout)
+
+    @property
+    def enabled(self) -> bool:
+        return self._enabled
+
+    @enabled.setter
+    def enabled(self, value: bool):
+        self._enabled = value
+
+    @property
+    def address(self) -> Address:
+        # bleak has no API for the address yet.
```

Comment:
I found a way to do it on Linux: https://github.com/adafruit/Adafruit_Blinka_bleio/blob/master/_bleio.py#L419-L423

Reply:
I am fetching the address using `bluetoothctl list` command instead. It's a bit simpler and the command clearly marks which adapter is the default, if there are multiple ones.